nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -2496,6 +2506,104 @@ mod test {
+    /// (multi-node) case.
+    #[test]
+    fn test_single_node_clickhouse() {
+        static TEST_NAME: &str = "test_single_node_clickhouse";
+        let logctx = test_setup_log(TEST_NAME);
+
+        // Use our example system as a starting point.
+        let (collection, input, blueprint1) = example(&logctx.log, TEST_NAME);
+
+        // We should not have any ClickHouse zones yet.
+        assert_eq!(
+            blueprint1
+                .all_omicron_zones(BlueprintZoneFilter::All)
+                .filter(|(_, zone)| zone.zone_type.is_clickhouse())
+                .count(),
+            0
+        );
+
+        // Construct a new blueprint that includes a single-node ClickHouse
+        // cluster on the first sled.
+        let mut blueprint_builder = BlueprintBuilder::new_based_on(
+            &logctx.log,
+            &blueprint1,
+            &input,
+            &collection,
+            TEST_NAME,
+        )
+        .expect("failed to build blueprint builder");
+        let sled_1 = blueprint_builder
+            .sled_ids_with_zones()
+            .next()
+            .expect("no first sled");
+        assert!(matches!(
+            blueprint_builder
+                .sled_ensure_zone_multiple_clickhouse(sled_1, 1)
+                .expect("can't add single-node ClickHouse zone to blueprint"),
+            EnsureMultiple::Changed { added: 1, removed: 0 }
+        ));
+
+        // Check that it's there.
+        let blueprint2 = blueprint_builder.build();
+        let diff = blueprint2.diff_since_blueprint(&blueprint1);
+        println!("1 -> 2 (added ClickHouse zone):\n{}", diff.display());
+        assert_eq!(
+            blueprint2
+                .all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)
+                .filter(|(_, zone)| zone.zone_type.is_clickhouse())
+                .count(),
+            1,
+            "can't find ClickHouse zone in new blueprint"
+        );
+
+        // Expunge the first sled and re-plan. This gets us an expunged
+        // ClickHouse zone, which the planner should immediately replace
+        // with one on another (non-expunged) sled.
+        let mut input_builder = input.into_builder();
+        input_builder
+            .sleds_mut()
+            .get_mut(&sled_1)
+            .expect("found sled 1 again")
+            .policy = SledPolicy::Expunged;
+        let input = input_builder.build();
+        let blueprint3 = Planner::new_based_on(
+            logctx.log.clone(),
+            &blueprint2,
+            &input,
+            "test_blueprint3",
+            &collection,
+        )
+        .expect("failed to create planner")
+        .with_rng_seed((TEST_NAME, "bp3"))
+        .plan()
+        .expect("failed to re-plan");
+
+        let diff = blueprint3.diff_since_blueprint(&blueprint2);
+        println!("2 -> 3 (expunged sled):\n{}", diff.display());
+        assert_eq!(
+            blueprint3
```

Comment:
Nit: Could probably also verify that the new zone is on a diff sled.

Reply:
Quite right, thanks! Done in 124059f76.

Reply:
Er, heads-up @andrewjstone: I had to revamp this test slightly to avoid breaking many other tests. `380efe4` adds a single-node ClickHouse zone to the example system, so that other tests won't have churn, but obviously this test has to start from that condition.